internal/providers/gce/gce.go

```diff
@@ -0,0 +1,118 @@
+// Copyright 2016 CoreOS, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// The gce provider fetches a remote configuration from the gce user-data
+// metadata service URL.
+
+package gce
+
+import (
+	"errors"
+	"net/http"
+	"strings"
+	"time"
+
+	"github.com/coreos/ignition/config"
+	"github.com/coreos/ignition/config/types"
+	"github.com/coreos/ignition/internal/log"
+	"github.com/coreos/ignition/internal/providers"
+	putil "github.com/coreos/ignition/internal/providers/util"
+	"github.com/coreos/ignition/internal/util"
+)
+
+const (
+	initialBackoff = 100 * time.Millisecond
+	maxBackoff     = 30 * time.Second
+	baseUrl        = "http://metadata.google.internal/computeMetadata/v1/instance/"
+	userdataUrl    = baseUrl + "attributes/user-data"
+	sshKeysUrl     = baseUrl + "attributes/ssh-keys"
+)
+
+var (
+	metadataHeader = http.Header{"Metadata-Flavor": []string{"Google"}}
+)
+
+type Creator struct{}
+
+func (Creator) Create(logger *log.Logger) providers.Provider {
+	return &provider{
+		logger:  logger,
+		backoff: initialBackoff,
+		client:  util.NewHttpClient(logger),
+	}
+}
+
+type provider struct {
+	logger    *log.Logger
+	backoff   time.Duration
+	client    util.HttpClient
+	rawConfig []byte
+}
+
+func (p provider) FetchConfig() (types.Config, error) {
+	cfg, err := config.Parse(p.rawConfig)
+	if err == nil || err == config.ErrEmpty || err == config.ErrDeprecated {
+		if keyCfg, err := p.fetchSSHKeys(); err == nil {
+			cfg = config.Append(cfg, keyCfg)
+		} else {
+			return cfg, err
+		}
+	}
+
+	return cfg, err
+}
+
+func (p *provider) IsOnline() bool {
+	p.rawConfig = p.client.FetchConfigWithHeader(userdataUrl, metadataHeader, http.StatusOK)
+	return (p.rawConfig != nil)
+}
+
+func (p provider) ShouldRetry() bool {
+	return true
+}
+
+func (p *provider) BackoffDuration() time.Duration {
+	return putil.ExpBackoff(&p.backoff, maxBackoff)
+}
+
+// fetchSSHKeys fetches and appends ssh keys to the config.
+func (p *provider) fetchSSHKeys() (types.Config, error) {
+	data, status, err := p.client.GetWithHeader(sshKeysUrl, metadataHeader)
+	if err != nil {
+		return types.Config{}, err
+	}
+	if status != http.StatusOK {
+		return types.Config{}, errors.New(http.StatusText(status))
+	}
+
+	keys := []string{}
+	for _, key := range strings.Split(string(data), "\n") {
+		tokens := strings.Split(key, ":")
```

Comment:
Use `SplitN(key, ":", 2)`. It'd be dumb to reject a key because ":" appeared in the comment trailing the key. The existing script is even more specific and only looks for ":" in the first word. https://github.com/coreos/coreos-overlay/blob/master/coreos-base/oem-gce/files/gce-ssh-key